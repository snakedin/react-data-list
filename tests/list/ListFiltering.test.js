import React from "react";
import { render, screen, waitForElementToBeRemoved, fireEvent } from "@testing-library/react";
import DataList, { Column } from "../../src/index";
import provider from "../mocks/providerMock";

it("DataList sorting", async () => {

    const { container } = render(
        <DataList provider={provider}>
            <Column id="id" label="ID"/>
            <Column id="name" />
            <Column id="year"/>
            <Column
                id="seasons"
                sort={false}
                filter={{
                    type: 'select',
                    items: [{ value: '', label: "" }, { value: 1, label: 1 }, { value: 2, label: 2}, { value: 5, label: 5 }]
                }}
            />
            <Column id="episodes" filter={false}/>
        </DataList>
    );

    // Default state
    expect(screen.getByRole("progressbar")).toBeInTheDocument();
    await waitForElementToBeRemoved(screen.getByRole("progressbar")).then(() => expect(container).toMatchSnapshot('1 without filters'));

    // Apply filter by ID
    fireEvent.change(screen.getAllByRole('textbox')[0], { target: { value: '18' }});
    fireEvent.blur(screen.getAllByRole('textbox')[0]);

    expect(screen.getByRole("progressbar")).toBeInTheDocument();
    await waitForElementToBeRemoved(screen.getByRole("progressbar")).then(() => expect(container).toMatchSnapshot('2 set filter by ID'));

    // Apply filter by seasons count
    fireEvent.change(screen.getAllByRole('textbox')[0], { target: { value: '' }});
    fireEvent.blur(screen.getAllByRole('textbox')[0]);

    await waitForElementToBeRemoved(screen.getByRole("progressbar"));
    fireEvent.change(screen.getAllByRole('combobox')[0], { target: { value: '5'}});

    expect(screen.getByRole("progressbar")).toBeInTheDocument();
    await waitForElementToBeRemoved(screen.getByRole("progressbar")).then(() => expect(container).toMatchSnapshot('3 set filter by seasons'));
});
