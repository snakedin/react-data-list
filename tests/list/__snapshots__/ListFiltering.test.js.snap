// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataList sorting: 1 without filters 1`] = `
<div>
  <div
    class="data-list"
  >
    <div>
      <table
        class="table table-striped table-hover table-bordered"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                ID
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                Name
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                Year
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm"
              >
                Seasons
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                Episodes
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <input
                class="form-control"
                type="text"
                value=""
              />
            </td>
            <td>
              <input
                class="form-control"
                type="text"
                value=""
              />
            </td>
            <td>
              <input
                class="form-control"
                type="text"
                value=""
              />
            </td>
            <td>
              <select
                class="form-control"
              >
                <option
                  value=""
                />
                <option
                  value="1"
                >
                  1
                </option>
                <option
                  value="2"
                >
                  2
                </option>
                <option
                  value="5"
                >
                  5
                </option>
              </select>
            </td>
            <td>
               
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              1
            </td>
            <td>
              Teenage Mutant Ninja Turtles
            </td>
            <td>
              1987
            </td>
            <td>
              10
            </td>
            <td>
              193
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              2
            </td>
            <td>
              Spider-Man
            </td>
            <td>
              1967
            </td>
            <td>
              3
            </td>
            <td>
              52
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              3
            </td>
            <td>
              She-Ra: Princess of Power
            </td>
            <td>
              1985
            </td>
            <td>
              2
            </td>
            <td>
              93
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              4
            </td>
            <td>
              The Pink Panther
            </td>
            <td>
              1992
            </td>
            <td>
              2
            </td>
            <td>
              60
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              5
            </td>
            <td>
              Biker Mice from Mars
            </td>
            <td>
              1993
            </td>
            <td>
              3
            </td>
            <td>
              65
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              6
            </td>
            <td>
              The Transformers
            </td>
            <td>
              1984
            </td>
            <td>
              4
            </td>
            <td>
              98
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              7
            </td>
            <td>
              The Flintstones
            </td>
            <td>
              1960
            </td>
            <td>
              6
            </td>
            <td>
              166
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              8
            </td>
            <td>
              The Spooktacular New Adventures of Casper
            </td>
            <td>
              1996
            </td>
            <td>
              4
            </td>
            <td>
              52
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              9
            </td>
            <td>
              Adventures of the Gummi Bears
            </td>
            <td>
              1985
            </td>
            <td>
              6
            </td>
            <td>
              65
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              10
            </td>
            <td>
              Visionaries: Knights of the Magical Light
            </td>
            <td>
              1987
            </td>
            <td>
              1
            </td>
            <td>
              13
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="row align-items-center mb-3 no-gutters"
      >
        <div
          class="col-12 col-lg-4 order-2 order-lg-1 text-left mb-3"
        >
          <ul
            class="pagination mb-0 justify-content-center justify-content-lg-start"
          >
            <li
              class="page-item disabled"
            >
              <button
                class="page-link"
                tabindex="-1"
              >
                <span
                  aria-hidden="true"
                >
                  «
                </span>
              </button>
            </li>
            <li
              class="page-item active"
            >
              <button
                class="page-link"
              >
                1
              </button>
            </li>
            <li
              class="page-item"
            >
              <button
                class="page-link"
              >
                2
              </button>
            </li>
            <li
              class="page-item"
            >
              <button
                class="page-link"
              >
                3
              </button>
            </li>
            <li
              class="page-item"
            >
              <button
                aria-label="Next"
                class="page-link"
              >
                <span
                  aria-hidden="true"
                >
                  »
                </span>
              </button>
            </li>
          </ul>
        </div>
        <div
          class="col-12 col-lg-4 order-1 order-lg-2 text-center mb-3"
        >
          <span>
            Showing 
            <b>
              1-10
            </b>
             of 
            <b>
              30
            </b>
             items.
          </span>
        </div>
        <div
          class="col-2 offset-5 col-lg-1 order-3 offset-lg-3 text-right mb-3"
        >
          <select
            class="form-control"
          >
            <option
              selected=""
              value="10"
            >
              10
            </option>
            <option
              value="20"
            >
              20
            </option>
            <option
              value="30"
            >
              30
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DataList sorting: 2 set filter by ID 1`] = `
<div>
  <div
    class="data-list"
  >
    <div>
      <table
        class="table table-striped table-hover table-bordered"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                ID
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                Name
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                Year
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm"
              >
                Seasons
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                Episodes
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <input
                class="form-control"
                type="text"
                value="18"
              />
            </td>
            <td>
              <input
                class="form-control"
                type="text"
                value=""
              />
            </td>
            <td>
              <input
                class="form-control"
                type="text"
                value=""
              />
            </td>
            <td>
              <select
                class="form-control"
              >
                <option
                  value=""
                />
                <option
                  value="1"
                >
                  1
                </option>
                <option
                  value="2"
                >
                  2
                </option>
                <option
                  value="5"
                >
                  5
                </option>
              </select>
            </td>
            <td>
               
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              18
            </td>
            <td>
              Ocean Girl
            </td>
            <td>
              1994
            </td>
            <td>
              4
            </td>
            <td>
              78
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="row align-items-center mb-3 no-gutters"
      >
        <div
          class="col-12 col-lg-4 order-2 order-lg-1 text-left mb-3"
        >
          <ul
            class="pagination mb-0 justify-content-center justify-content-lg-start"
          />
        </div>
        <div
          class="col-12 col-lg-4 order-1 order-lg-2 text-center mb-3"
        >
          <span>
            Showing 
            <b>
              1-1
            </b>
             of 
            <b>
              1
            </b>
             items.
          </span>
        </div>
        <div
          class="col-2 offset-5 col-lg-1 order-3 offset-lg-3 text-right mb-3"
        >
          <select
            class="form-control"
          >
            <option
              selected=""
              value="10"
            >
              10
            </option>
            <option
              value="20"
            >
              20
            </option>
            <option
              value="30"
            >
              30
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DataList sorting: 3 set filter by seasons 1`] = `
<div>
  <div
    class="data-list"
  >
    <div>
      <table
        class="table table-striped table-hover table-bordered"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                ID
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                Name
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                Year
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm"
              >
                Seasons
              </span>
            </th>
            <th
              scope="col"
            >
              <span
                class="btn btn-sm btn-link asc"
              >
                Episodes
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <input
                class="form-control"
                type="text"
                value=""
              />
            </td>
            <td>
              <input
                class="form-control"
                type="text"
                value=""
              />
            </td>
            <td>
              <input
                class="form-control"
                type="text"
                value=""
              />
            </td>
            <td>
              <select
                class="form-control"
              >
                <option
                  value=""
                />
                <option
                  value="1"
                >
                  1
                </option>
                <option
                  value="2"
                >
                  2
                </option>
                <option
                  selected=""
                  value="5"
                >
                  5
                </option>
              </select>
            </td>
            <td>
               
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              13
            </td>
            <td>
              X-Men
            </td>
            <td>
              1992
            </td>
            <td>
              5
            </td>
            <td>
              76
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              24
            </td>
            <td>
              The Raccoons
            </td>
            <td>
              1980
            </td>
            <td>
              5
            </td>
            <td>
              1960
            </td>
          </tr>
          <tr
            class=""
            role="data-list-row"
          >
            <td>
              27
            </td>
            <td>
              Hey Arnold!
            </td>
            <td>
              1996
            </td>
            <td>
              5
            </td>
            <td>
              100
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="row align-items-center mb-3 no-gutters"
      >
        <div
          class="col-12 col-lg-4 order-2 order-lg-1 text-left mb-3"
        >
          <ul
            class="pagination mb-0 justify-content-center justify-content-lg-start"
          />
        </div>
        <div
          class="col-12 col-lg-4 order-1 order-lg-2 text-center mb-3"
        >
          <span>
            Showing 
            <b>
              1-3
            </b>
             of 
            <b>
              3
            </b>
             items.
          </span>
        </div>
        <div
          class="col-2 offset-5 col-lg-1 order-3 offset-lg-3 text-right mb-3"
        >
          <select
            class="form-control"
          >
            <option
              selected=""
              value="10"
            >
              10
            </option>
            <option
              value="20"
            >
              20
            </option>
            <option
              value="30"
            >
              30
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
</div>
`;
